File: src/agents/critic_agent.py
"""Critic agent that validates delegation/tool-use policy compliance."""

from __future__ import annotations

from google.adk import Agent

DEFAULT_CRITIC_AGENT_PROMPT = (
    "You are CriticAgent. Validate whether the candidate answer follows delegation policy.\n"
    "Policy:\n"
    "- Math-like queries MUST involve MathAgent handoff/tool usage evidence.\n"
    "- Factual/latest/source-seeking queries MUST involve ResearchAgent handoff and web-search evidence.\n"
    "- If policy-triggering query was answered directly without required delegation, reject.\n"
    "Return JSON ONLY with schema:\n"
    '{"compliant": true|false, "required_action": "accept"|"delegate_research"|"delegate_math"|"retry_with_instruction", "rationale": "short explanation"}\n'
    "No markdown or extra keys."
)


def get_critic_agent(
    system_prompt: str | None = None,
    model: str = "gemini-2.0-flash-lite",
) -> Agent:
    """Create the critic agent."""
    prompt = system_prompt if system_prompt is not None else DEFAULT_CRITIC_AGENT_PROMPT
    return Agent(
        name="CriticAgent",
        model=model,
        instruction=prompt,
        tools=[],
    )

File: src-ts/agents/critic-agent.ts
import { openaiModel } from "../model.js";
import { getAISDK } from "../tracing.js";
import type { SerializedMessage } from "../types.js";

export const DEFAULT_CRITIC_AGENT_PROMPT = `You are CriticAgent. Validate whether the candidate answer follows delegation policy.
Policy:
- Math-like queries MUST involve MathAgent handoff/tool usage evidence.
- Factual/latest/source-seeking queries MUST involve ResearchAgent handoff and web-search evidence.
- If policy-triggering query was answered directly without required delegation, reject.
Return JSON ONLY with schema:
{"compliant": true|false, "required_action": "accept"|"delegate_research"|"delegate_math"|"retry_with_instruction", "rationale": "short explanation"}
No markdown or extra keys.`;

export type CriticDecision = {
  compliant: boolean;
  required_action: "accept" | "delegate_research" | "delegate_math" | "retry_with_instruction";
  rationale: string;
};

export async function runCriticAgent(options: {
  query: string;
  candidateFinalOutput: string;
  messages: SerializedMessage[];
  model?: string;
  systemPrompt?: string;
}): Promise<Partial<CriticDecision>> {
  const aiSdk = getAISDK();
  const payload = {
    query: options.query,
    candidate_final_output: options.candidateFinalOutput,
    messages: options.messages,
  };

  const response = await aiSdk.generateText({
    model: openaiModel(options.model ?? "gpt-4.1-mini"),
    system: options.systemPrompt ?? DEFAULT_CRITIC_AGENT_PROMPT,
    prompt: `Evaluate this candidate against delegation policy and return JSON only.\n${JSON.stringify(payload)}`,
  });

  const rawText = (response.text ?? "").trim();
  try {
    return JSON.parse(rawText) as Partial<CriticDecision>;
  } catch {
    return {};
  }
}
